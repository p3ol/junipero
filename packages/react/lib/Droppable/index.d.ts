import { ComponentPropsWithoutRef, ReactNode } from 'react';

declare interface DroppableProps extends ComponentPropsWithoutRef<any> {
  className?: string;
  children?: ReactNode | JSX.Element;
  disabled?: boolean;
  onDrop?(data: any, direction: string, e: Event): void;
  onDragOver?(e: Event, direction: string): void;
  onDragLeave?(e: Event): void;
}

declare function Droppable(props: DroppableProps):
  ReactNode | JSX.Element;

export default Droppable;
