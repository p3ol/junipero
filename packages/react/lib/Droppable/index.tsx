import {
  type ComponentPropsWithRef,
  type DragEvent,
  Children,
  cloneElement,
  forwardRef,
  useState,
  useEffect,
} from 'react';
import { type ForwardedProps, classNames } from '@junipero/core';
import PropTypes from 'prop-types';

export declare type DraggingPositionType = 'before' | 'after';

export declare interface DroppableProps extends ComponentPropsWithRef<any> {
  className?: string;
  children?: JSX.Element;
  disabled?: boolean;
  onDrop?(data: any, direction: DraggingPositionType, e: DragEvent): void;
  onDragOver?(e: DragEvent, direction: DraggingPositionType): void;
  onDragLeave?(e: DragEvent): void;
}

const Droppable = forwardRef((props: DroppableProps, ref) => {
  const [dragging, setDragging] = useState(false);
  const [stack, setStack] = useState(0);
  const [draggingPos, setDraggingPos] = useState(null);
  const {
    className,
    children,
    disabled = false,
    onDrop,
    onDragOver,
    onDragLeave,
    ...rest
  } = props;

  useEffect(() => {
    if (stack <= 0) {
      setDragging(false);
    }
  }, [stack]);

  const onDragEnter_ = () => {
    if (disabled) {
      return;
    }

    setStack(s => s + 1);
    setDragging(true);
  };

  const onDragLeave_ = (e: any) => {
    if (disabled) {
      return;
    }

    setStack(s => s - 1);
    setDraggingPos(null);
    onDragLeave?.(e);
  };

  const onDrop_ = (e: DragEvent) => {
    if (disabled) {
      return;
    }

    setStack(0);
    setDragging(false);
    onDrop?.(JSON.parse(e.dataTransfer.getData('text')), draggingPos, e);
    e.preventDefault();
    setDraggingPos(null);
  };

  const onDragOver_ = (e: DragEvent) => {
    if (disabled) {
      return;
    }

    const targetRect = e.currentTarget.getBoundingClientRect();
    const targetMiddleY = targetRect?.top + targetRect?.height / 2;
    let draggingPosition;

    if (e.clientY >= targetMiddleY) {
      draggingPosition = 'after';
    } else if (e.clientY < targetMiddleY) {
      draggingPosition = 'before';
    }

    setDraggingPos(draggingPosition);
    e.preventDefault();
    onDragOver?.(e, draggingPosition as DraggingPositionType);

    return false;
  };

  const child = Children.only(children);

  return cloneElement(child, {
    ...rest,
    ref,
    className: classNames(
      className,
      child.props.className,
      {
        'drag-enter': !disabled && dragging,
        'drag-top': !disabled && dragging && draggingPos &&
          draggingPos === 'before',
        'drag-bottom': !disabled && dragging && draggingPos &&
          draggingPos === 'after',
      },
    ),
    onDragEnter: onDragEnter_,
    onDragLeave: onDragLeave_,
    onDrop: onDrop_,
    onDragOver: onDragOver_,
  });
}) as ForwardedProps<DroppableProps, any>;

Droppable.displayName = 'Droppable';
Droppable.propTypes = {
  disabled: PropTypes.bool,
  onDrop: PropTypes.func,
  onDragOver: PropTypes.func,
  onDragLeave: PropTypes.func,
};

export default Droppable;
