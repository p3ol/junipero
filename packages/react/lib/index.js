// Display
export { default as Abstract } from './Abstract';
export { default as Alert } from './Alert';
export { default as Alerts } from './Alerts';
export { default as AlertsControl } from './AlertsControl';
export { default as Badge } from './Badge';
export { default as BreadCrumb } from './BreadCrumb';
export { default as BreadCrumbItem } from './BreadCrumbItem';
export { default as Button } from './Button';
export { default as Calendar } from './Calendar';
export { default as Card } from './Card';
export { default as Draggable } from './Draggable';
export { default as Dropdown } from './Dropdown';
export { default as DropdownGroup } from './DropdownGroup';
export { default as DropdownItem } from './DropdownItem';
export { default as DropdownMenu } from './DropdownMenu';
export { default as DropdownToggle } from './DropdownToggle';
export { default as Droppable } from './Droppable';
export { default as Label } from './Label';
export { default as List } from './List';
export { default as ListCell } from './ListCell';
export { default as ListColumn } from './ListColumn';
export { default as ListItem } from './ListItem';
export { default as Modal } from './Modal';
export { default as Slider } from './Slider';
export { default as Spinner } from './Spinner';
export { default as Tab } from './Tab';
export { default as Tabs } from './Tabs';
export { default as Tag } from './Tag';
export { default as Toast } from './Toast';
export { default as Toasts } from './Toasts';
export { default as ToastsControl } from './ToastsControl';
export { default as Toggle } from './Toggle';
export { default as Tooltip } from './Tooltip';
export { default as TouchableZone } from './TouchableZone';
export { default as Transition } from './Transition';

// Forms
export { default as CheckboxField } from './CheckboxField';
export { default as CodeField } from './CodeField';
export { default as ColorField } from './ColorField';
export { default as DateField } from './DateField';
export { default as FieldAddon } from './FieldAddon';
export { default as FieldControl } from './FieldControl';
export { default as FieldGroup } from './FieldGroup';
export { default as RadioField } from './RadioField';
export { default as SelectField } from './SelectField';
export { default as TextField } from './TextField';

// Logos
export {
  PooolLogo,
  PooolCompactLogo,
  AccessLogo,
  SubscribeLogo,
  ConnectLogo,
  EngageLogo,
  FlowLogo,
} from './Logo';

// Core
export {
  classNames,
  addClass,
  removeClass,
  mockState,
  isUndefined,
  isNull,
  isArray,
  isObject,
  isDate,
  exists,
  get,
  set,
  omitBy,
  omit,
  pick,
  cloneDeep,
  fromPairs,
  mergeDeep,
  ensureNode,
  findDeep,
  filterDeep,
  ensureMinMax,
  getFloatPrecision,
  formatThousand,
} from '@junipero/core';

// Hooks
export {
  useEventListener,
  useInterval,
  useTimeout,
  useEffectAfterMount,
  useLayoutEffectAfterMount,
} from '@junipero/hooks';

export {
  useAlerts,
  useToasts,
} from './hooks';

// Icons
export {
  Arrows,
  Remove,
  ArrowLeft,
  ArrowRight,
  ArrowDown,
  ArrowUp,
  Time,
  Check,
} from './icons';
