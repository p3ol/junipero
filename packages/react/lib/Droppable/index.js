import { Children, cloneElement, forwardRef, useState, useEffect } from 'react';
import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';

const Droppable = forwardRef(({
  className,
  children,
  disabled = false,
  onDrop,
  onDragOver,
  onDragLeave,
  ...rest
}, ref) => {
  const [dragging, setDragging] = useState(false);
  const [stack, setStack] = useState(0);
  const [draggingPos, setDraggingPos] = useState(null);

  useEffect(() => {
    if (stack <= 0) {
      setDragging(false);
    }
  }, [stack]);

  const onDragEnter_ = () => {
    if (disabled) {
      return;
    }

    setStack(s => s + 1);
    setDragging(true);
  };

  const onDragLeave_ = e => {
    if (disabled) {
      return;
    }

    setStack(s => s - 1);
    setDraggingPos(null);
    onDragLeave?.(e);
  };

  const onDrop_ = e => {
    if (disabled) {
      return;
    }

    setStack(0);
    setDragging(false);
    onDrop?.(JSON.parse(e.dataTransfer.getData('text')), draggingPos, e);
    e.preventDefault();
    setDraggingPos(null);
  };

  const onDragOver_ = e => {
    if (disabled) {
      return;
    }

    const targetRect = e.currentTarget.getBoundingClientRect();
    const targetMiddleY = targetRect?.top + targetRect?.height / 2;
    let draggingPosition;

    if (e.clientY >= targetMiddleY) {
      draggingPosition = 'after';
    } else if (e.clientY < targetMiddleY) {
      draggingPosition = 'before';
    }

    setDraggingPos(draggingPosition);
    e.preventDefault();
    onDragOver?.(e, draggingPosition);

    return false;
  };

  const child = Children.only(children);

  return cloneElement(child, {
    ...rest,
    ref,
    className: classNames(
      className,
      child.props.className,
      {
        'drag-enter': !disabled && dragging,
        'drag-top': !disabled && dragging && draggingPos &&
          draggingPos === 'before',
        'drag-bottom': !disabled && dragging && draggingPos &&
          draggingPos === 'after',
      },
    ),
    onDragEnter: onDragEnter_,
    onDragLeave: onDragLeave_,
    onDrop: onDrop_,
    onDragOver: onDragOver_,
  });
});

Droppable.displayName = 'Droppable';
Droppable.propTypes = {
  disabled: PropTypes.bool,
  onDrop: PropTypes.func,
  onDragOver: PropTypes.func,
  onDragLeave: PropTypes.func,
};

export default Droppable;
