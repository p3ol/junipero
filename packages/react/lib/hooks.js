import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
