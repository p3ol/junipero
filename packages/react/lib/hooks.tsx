import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
