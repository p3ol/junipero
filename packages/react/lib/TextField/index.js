import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
