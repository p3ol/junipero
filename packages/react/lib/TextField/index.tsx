import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
