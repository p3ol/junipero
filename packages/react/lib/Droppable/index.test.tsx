import { createEvent, fireEvent, render } from '@testing-library/react';

import Droppable from '.';

describe('<Droppable />', () => {
  it('should render', () => {
    const onDragOver = jest.fn();
    const onDragLeave = jest.fn();
    const { container, unmount } = render(
      <Droppable onDragOver={onDragOver} onDragLeave={onDragLeave}>
        <p>hello</p>
      </Droppable>
    );
    expect(container).toMatchSnapshot();

    fireEvent.dragEnter(container.querySelector('p'));
    expect(container).toMatchSnapshot('drag entered');

    const dragAboveEvent = createEvent.dragOver(container.querySelector('p'));
    Object.defineProperty(dragAboveEvent, 'clientY', { value: -1 });
    fireEvent(container.querySelector('p'), dragAboveEvent);
    expect(onDragOver).toHaveBeenCalled();
    expect(container).toMatchSnapshot('drag above');

    const dragBelowEvent = createEvent.dragOver(container.querySelector('p'));
    Object.defineProperty(dragBelowEvent, 'clientY', { value: 1 });
    fireEvent(container.querySelector('p'), dragBelowEvent);
    expect(container).toMatchSnapshot('drag below');

    fireEvent.dragLeave(container.querySelector('p'));
    expect(container).toMatchSnapshot('drag leave');
    expect(onDragLeave).toHaveBeenCalled();

    unmount();
  });

  it('should trigger onDrop event when drop', () => {
    const onDrop = jest.fn();
    const { container, unmount } = render(
      <Droppable onDrop={onDrop}><p>hello</p></Droppable>
    );

    fireEvent.dragEnter(container.querySelector('p'));
    const dragOverEvent = createEvent.dragOver(container.querySelector('p'));
    Object.defineProperty(dragOverEvent, 'clientY', { value: 1 });
    fireEvent(container.querySelector('p'), dragOverEvent);
    expect(container).toMatchSnapshot('drag over');

    fireEvent.drop(container.querySelector('p'), {
      dataTransfer: { getData: () => '{}' },
    });

    expect(onDrop).toHaveBeenCalled();

    unmount();
  });

  it('should not trigger anything when disabled', () => {
    const onDrop = jest.fn();
    const onDragOver = jest.fn();
    const onDragLeave = jest.fn();
    const { container, unmount } = render(
      <Droppable
        onDrop={onDrop}
        onDragOver={onDragOver}
        onDragLeave={onDragLeave}
        disabled
      >
        <p>hello</p>
      </Droppable>
    );

    fireEvent.dragEnter(container.querySelector('p'));
    const dragOverEvent = createEvent.dragOver(container.querySelector('p'));
    Object.defineProperty(dragOverEvent, 'clientY', { value: 1 });
    fireEvent(container.querySelector('p'), dragOverEvent);
    expect(container).toMatchSnapshot('drag over');

    fireEvent.drop(container.querySelector('p'), {
      dataTransfer: { getData: () => '{}' },
    });

    expect(onDrop).not.toHaveBeenCalled();
    expect(onDragOver).not.toHaveBeenCalled();
    expect(onDragLeave).not.toHaveBeenCalled();

    unmount();
  });
});
