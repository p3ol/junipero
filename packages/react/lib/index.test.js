import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
