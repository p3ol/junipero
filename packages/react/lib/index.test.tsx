import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
