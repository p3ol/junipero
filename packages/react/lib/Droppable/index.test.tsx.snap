// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Droppable /> > should not trigger anything when disabled > drag over 1`] = `
<div>
  <p
    class=""
  >
    hello
  </p>
</div>
`;

exports[`<Droppable /> > should render > drag above 1`] = `
<div>
  <p
    class="drag-enter drag-top"
  >
    hello
  </p>
</div>
`;

exports[`<Droppable /> > should render > drag below 1`] = `
<div>
  <p
    class="drag-enter drag-bottom"
  >
    hello
  </p>
</div>
`;

exports[`<Droppable /> > should render > drag entered 1`] = `
<div>
  <p
    class="drag-enter"
  >
    hello
  </p>
</div>
`;

exports[`<Droppable /> > should render > drag leave 1`] = `
<div>
  <p
    class=""
  >
    hello
  </p>
</div>
`;

exports[`<Droppable /> > should render 1`] = `
<div>
  <p
    class=""
  >
    hello
  </p>
</div>
`;

exports[`<Droppable /> > should trigger onDrop event when drop > drag over 1`] = `
<div>
  <p
    class="drag-enter drag-bottom"
  >
    hello
  </p>
</div>
`;
